test/uploaded_file_test.rb

```diff
@@ -319,6 +319,23 @@ def open(id)
       assert_match "file", uploaded_file.download.read
     end
 
+    it "deletes the Tempfile if there's an error in downloading the file (Shrine class level)" do
+      io_error = ->(src, dest) { raise IOError }
+      mock = Minitest::Mock.new
+      mock.expect(:close!, nil)
+      def mock.path
+        "some_path"
+      end
+      @uploader.storage.instance_eval { undef download }
+      uploaded_file = @uploader.upload(fakeio("file"))
+      Tempfile.stub(:new, mock) do
+        IO.stub(:copy_stream, io_error) do
+          assert_raises(IOError) { uploaded_file.download }
+        end
+      end
+      mock.verify
+    end
+
```

Comment:
Instead of creating isolated mocks, what do you think about stubbing with actual `Tempfile` objects? I generally prefer that because it avoids me to use method call expectations. I would also prefer if we stubbed `Storage#open` to return an exception instead of `IO.copy_stream`, so that tests communicate what we're actually guarding against; `IO.copy_stream` will fail only if the source object is not a real IO, so for me that's not a concern as then it will consistently fail, but `Storage#open` can fail with temporary network exceptions, and that's why I wanted to guard against.

So I was thinking something in the lines of:

```rb
@uploader.storage.instance_eval { undef download }
@uploader.storage.instance_eval { def open(id); raise SystemCallError, "error occurred"; end }
uploaded_file = @uploader.upload(fakeio)
Tempfile.stubs(:new, tempfile = Tempfile.new) do
  assert_raises(SystemCallError) { uploaded_file.download }
end
assert tempfile.closed?
assert_nil tempfile.path # when tempfile gets deleted its #path becomes nil
```

Reply:
You could simplify this even more by creating a `Shrine::UploadedFile` object that doesn't exist (you can use the `#uploaded_file` helper method that already exists on top of the test file), and then `Shrine::Storage::Memory` (which is a superclass of the `Shrine::Storage::Test` we're working with here) will raise a `KeyError` because it uses `Hash#fetch` in `#open`.

```rb
@uploader.storage.instance_eval { undef download }
Tempfile.stubs(:new, tempfile = Tempfile.new) do
  assert_raises(KeyError) { uploaded_file.download }
end
assert tempfile.closed?
assert_nil tempfile.path # when tempfile gets deleted its #path becomes nil
```

Reply:
I like your approach of stubbing with actual `Tempfile` better. It also reduces having to create other methods in the stub so the code can run up to the target code of interest (e.g. having to add Tempfile#path).

I understand regarding `Storage#open` - makes sense. Will make the changes.